### DISCUSSION TITLE
Strongswan Android VPN app with PQC support not working

### DISCUSSION THREAD

**Discussion by Dhanasree234** [strongswan/strongswan#2344] [Q&A]
I am using the Android strong-swan VPN client app version 2.5.1 and the strong-swan library version 6.0.0 beta6. I am trying to enable Post-Quantum Cryptography (PQC) in the strong-swan VPN client app. I am not able to establish a connection between the strongswan server and the strongswan client. The strongswan VPN server has PQC support and it has been tested with strongswan VPN linux client. Here is a detailed description of my setup and the steps I have taken. I have also attached the swanctl.conf and strongswan.conf files below: 

[swanctl.txt](https://github.com/user-attachments/files/16188601/swanctl.txt)
[strongswan.txt](https://github.com/user-attachments/files/16188602/strongswan.txt)

### Setup

1. **strong-swan Versions:**
	   - Android strong-swan app version: 2.5.1
	   - strong-swan library version: 6.0.0 beta6
	   
2. **liboqs Version:**
	   - liboqs version: 0.8.0

3. **Android Architectures:**
	   - Built the liboqs library and include folder for different Android architectures 	(armeabi-v7a, arm64-v8a, x86, x86_64).
	   - Placed the compiled liboqs library in the specified path within the Android project.

#### 1. `Android.mk` in the `jni` Folder
Added the following configuration for loading the `oqs` plugin:

	include $(CLEAR_VARS)
	LOCAL_MODULE := liboqs
	LOCAL_SRC_FILES := $(LOCAL_PATH)/liboqs/$(TARGET_
	LOCAL_EXPORT_C_INCLUDES := $(LOCAL_PATH)/liboqs/$(TARGET_
	include $(PREBUILT_SHARED_LIBRARY)

2. Linking liboqs.so in Android.mk of the libstrongswan Folder
Added the following configuration for linking the liboqs.so library:

	LOCAL_SRC_FILES += $(call add_plugin, oqs)
	#ifneq ($(call plugin_enabled, oqs),)
	LOCAL_SHARED_LIBRARIES += liboqs

### Log:

 **Strongswan vpn client android app  Logs :**    

Jul 10 18:45:43 00[DMN] +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-
Jul 10 18:45:43 00[DMN] Starting IKE service (strongSwan , Android 10 - build-20201105005518/2020-10-
Jul 10 18:45:43 00[LIB] loaded plugins: androidbridge charon android-log socket-default openssl nonce oqs pkcs1 pem x509 xcbc kdf revocation eap-identity eap-mschapv2 eap-md5 eap-gtc eap-tls
Jul 10 18:45:43 00[JOB] spawning 16 worker threads
Jul 10 18:45:43 06[IKE] initiating IKE_SA android[24] to 192.168.1.100
Jul 10 18:45:43 06[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) N(IKE_INT_SUP) ]
Jul 10 18:45:43 06[NET] sending packet: from 192.168.1.101[57137] to 192.168.1.100[500] (256 bytes)
Jul 10 18:45:43 07[NET] received packet: from 192.168.1.100[500] to 192.168.1.101[57137] (36 bytes)
**Jul 10 18:45:43 07[ENC] parsed IKE_SA_INIT response 0 [ N(NO_PROP) ]
Jul 10 18:45:43 07[IKE] received NO_PROPOSAL_CHOSEN notify error**

**Strongswan server Logs :** 

07[CFG] added vici connection: rw
00[DMN] conns: loaded connection 'rw'
00[DMN] conns: successfully loaded 1 connections, 0 unloaded
00[DMN] executing start script 'pools' (/opt/build/sbin/swanctl --load-pools)
no files found matching '/opt/build/etc/swanctl/conf.
08[CFG] added vici pool rw_pool: 192.168.100.1, 100 entries
00[DMN] pools: loaded pool 'rw_pool'
00[DMN] pools: successfully loaded 1 pools, 0 unloaded
12[NET] received packet: from 192.168.1.101[44794] to 192.168.1.100[500] (256 bytes)
12[ENC] parsed IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) N(IKE_INT_SUP) ]
12[IKE] 192.168.1.101 is initiating an IKE_SA
**12[CFG] an algorithm from private space would match, but peer implementation is unknown, skipped**
12[CFG] received proposals: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/CURVE_25519/KE1_KYBER_L3
12[CFG] configured proposals: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/CURVE_25519/KE1_KYBER_L3
12[IKE] remote host is behind NAT
**12[IKE] received proposals unacceptable
12[ENC] generating IKE_SA_INIT response 0 [ N(NO_PROP) ]**
12[NET] sending packet: from 192.168.1.100[500] to 192.168.1.101[44794] (36 bytes)
14[NET] received packet: from 192.168.1.101[57137] to 192.168.1.100[500] (256 bytes)
14[ENC] parsed IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) N(IKE_INT_SUP) ]

**Comment by tobiasbrunner** [comment]
> **12[CFG] an algorithm from private space would match, but peer implementation is unknown, skipped**

Either the client has to send the strongSwan vendor ID (can be enabled via `charon.send_vendor_id`), or the server has to be configured to accept private identifiers without having received that vendor ID, which can be enabled via `charon.accept_private_algs`.

**Comment by Dhanasree234** [comment]
Thanks for replying sir
I enabled charon.accept_private_algs at the server and now it went ahead and stuck with a different error on the strongswan android vpn client as shown below :

Jul 12 18:23:02 05[CFG] selected proposal: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/CURVE_25519/KE1_KYBER_L3
Jul 12 18:23:02 05[IKE] faking NAT situation to enforce UDP encapsulation
**Jul 12 18:23:02 05[LIB] OQS RNG could not be switched to openssl
Jul 12 18:23:02 05[IKE] negotiated key exchange method KYBER_L3 not supported**

pls suggest. thanks in advance

**Reply by tobiasbrunner** [reply]
Seems to be an error due to `OQS_randombytes_switch_algorithm()` failing. Since that's a function provided by `liboqs`, I can't help you.

**Comment by Dhanasree234** [comment]
Thanks for replying sir .
I could proceed further by enabling OPENSSL_USE_OPTION=ON in the liboqs library but currently I am encounterong a new error on the strongswan vpn android client app as shown below :

 Jul 16 12:32:48 07[IKE] server requested EAP_MSCHAPV2 authentication (id 0x98)
Jul 16 12:32:48 07[IKE] EAP-MS-CHAPv2 generating NT-Response failed
Jul 16 12:32:48 07[IKE] EAP_MSCHAPV2 method failed
Jul 16 12:32:48 07[ENC] generating INFORMATIONAL request 3 [ N(AUTH_FAILED) ]
Jul 16 12:32:48 07[NET] sending packet: from 192.168.1.101[54814] to 192.168.1.100[4500] (80 bytes)

Please suggest..

**Reply by tobiasbrunner** [reply]
See #2351.
